src/mqtt_interface.rs

```diff
@@ -0,0 +1,258 @@
+    Ok(string)
+}
+
+/// An interface for managing MQTT settings.
+pub struct MqttInterface<T, S>
+where
+    T: StringSet,
+    S: TcpStack,
+{
+    // TODO: Allow the user to specify buffer size.
+    client: Option<MqttClient<minimq::consts::U256, S>>,
+    pub settings: T,
+
+    subscribed: bool,
+    settings_topic: String<consts::U128>,
+    commit_topic: String<consts::U128>,
+    default_response_topic: String<consts::U128>,
+    id: String<consts::U128>,
+}
+
+impl<T, S> MqttInterface<T, S>
+where
+    T: StringSet,
+    S: TcpStack,
+{
+    /// Construct a new settings interface using the network stack.
+    ///
+    /// # Args
+    /// * `stack` - The TCP network stack to use for communication.
+    /// * `id` - The ID for uniquely identifying the device.
+    /// * `broker` - The IpAddress of the MQTT broker.
+    /// * `settings` - The initial settings of the interface.
+    ///
+    /// # Returns
+    /// A new `MqttInterface` object that can be used for settings configuration and telemtry.
+    pub fn new<'a>(
+        stack: S,
+        id: &'a str,
+        broker: IpAddr,
+        settings: T,
+    ) -> Result<Self, Error<S::Error>> {
+        let client: MqttClient<minimq::consts::U256, _> = MqttClient::new(broker, id, stack)?;
+
+        let settings_topic = generate_topic(id, "settings/#").or(Err(Error::IdTooLong))?;
+        let commit_topic = generate_topic(id, "commit").or(Err(Error::IdTooLong))?;
+        let default_response_topic = generate_topic(id, "log").or(Err(Error::IdTooLong))?;
+
+        Ok(Self {
+            client: Some(client),
+            subscribed: false,
+            settings,
+
+            settings_topic,
+            default_response_topic,
+            commit_topic,
+
+            // Note(unwrap): We can safely assume the ID is less than our storage size, since we
+            // generate longer strings above.
+            id: String::from(id),
+        })
+    }
+
+    /// Called to periodically service the MQTT telemetry interface.
+    ///
+    /// # Note
+    /// This function should be called whenever the underlying network stack has processed incoming
+    /// or outgoing data.
+    ///
+    /// # Returns
+    /// An `Action` indicating what action should be taken by the user application.
+    pub fn update(&mut self) -> Result<Action, Error<S::Error>> {
+        // Note(unwrap): We maintain strict control of the client object, so it should always be
+        // present.
+        let mut client = self.client.take().unwrap();
+
+        // If we are not yet subscribed to the necessary topics, subscribe now.
+        if !self.subscribed && client.is_connected()? {
+            client.subscribe(&self.settings_topic, &[])?;
+            client.subscribe(&self.commit_topic, &[])?;
+            self.subscribed = true;
+        }
+
+        let mut result = Action::Continue;
+
+        let result = match client.poll(|client, topic, message, properties| {
```

Comment:
This is kinda confusing have multiple `result` here, especially because the first `result` is used inside the match statement that calculates the second `result`.

I wonder if we could make the logic of this match statement more clear by creating a separate function or something, it's nearly 100 lines.